### DISCUSSION TITLE
Interpreting HMM results with covariates

### DISCUSSION THREAD

**Discussion by twangss** [atsa-es/fish550-2023#37] [General]
I'm having a hard time understanding the output of depmix with covariates. I provided the output below. My interpretation is that SST decreases the probability of transitioning from state 1 to state 2, while Deep_temp increases the prob. Is the first batch of coeff for state 1 to state 2, while the second batch is state 2 to state 1? Also the initial state probabilities of 1 and 0 seems incorrect? Without the covariates, the transition probabilities are not so binary. Are these interpretations on the right track? Thanks!

Initial state probabilities model 
pr1 pr2 
  1   0 

Transition model for state (component) 1 
Model of type multinomial (mlogit), formula: ~SST + Deep_temperature
Coefficients: 
                 St1       St2
(Intercept)        0  123.1808
SST                0 -104.5492
Deep_temperature   0  161.8352
Probalities at zero values of the covariates.
3.186094e-54 1 

Transition model for state (component) 2 
Model of type multinomial (mlogit), formula: ~SST + Deep_temperature
Coefficients: 
                 St1         St2
(Intercept)        0  5.54345018
SST                0 -0.33007916
Deep_temperature   0  0.02560386
Probalities at zero values of the covariates.
0.003897751 0.9961022

**Comment by ericward-noaa** [comment]
Great questions -- I think it's worth me going through some of this today in class, because I'm sure you're not the only one who's run into these issues. I agree these mlogit models are super confusing. I'm going to try to step through this -- but only focus on 1 covariate. To start from the models we talked about in class, we start with the intercept only model / no covariates

```
data(speed)
mod <- depmix(list(rt~1),
              data=speed,
              nstates=2,
              family=list(gaussian()))
fit <- fit(mod)
```

you can then look at 
`summary(fit)`

 or
```
fit@response
fit@transition
```

To see the estimated intercepts for each of the 2 states

**Comment by ericward-noaa** [comment]
Using the same dataset, we'll now add in covariates on the intercept:

```
mod <- depmix(list(rt~corr),
              data=speed,
              nstates=2,
              family=list(gaussian()))
fit <- fit(mod)
```

In this case, fit@transition is the same, but
`fit@response`
gives 2 intercepts / slopes / sds. In other words, there are 2 regression models being fit -- either in state 1 or state 2

**Comment by ericward-noaa** [comment]
Now, we can add in covariates onto the transition probabilities. This is where things get trickier.

```
mod <- depmix(list(rt~1),
              transition=~corr,
              data=speed,
              nstates=2,
              family=list(gaussian()))
fit <- fit(mod)
```

`fit@response` 
contains 2 intercepts, 2 sds as expected
`attributes(fit)$transition` 
gives the intercept and slope for the transition probabilities for each model in logit space. Why are some set to 0 arbitrarily? These are the reference cases that we'll see below -- probabilities are normalized, so these aren't identifiable.

To look at the effect of covariates, if often helps to look at several levels of the covariate. I put together this little function for this example (2 states, 1 covariate on transition probabilities). Note that the steps are (1) create transition matrix in logit space, and (2) back transform

```
convert_probs = function(fit, coef_value = 0) {
  # extract coefficients
  p1 <- fit@transition[[1]]@parameters$coefficients
  p2 <- fit@transition[[2]]@parameters$coefficients
  
  # assemble transition matrix in mlogit space
  m <- matrix(0, 2, 2)
  m[1,] <- p1[1,] + p1[2,] * coef_value
  m[2,] <- p2[1,] + p2[2,] * coef_value
  
  # exponentiate
  m <- exp(m)
  
  # normalize
  m[1,] <- m[1,] / sum(m[1,])
  m[2,] <- m[2,] / sum(m[2,])
  return(m)
}
```

So to look at the effects of the matrix when covariate 'corr' is 0,

`convert_probs(fit, coef_value = 0)`

and compare these to

`convert_probs(fit, coef_value = 1)`

**Comment by ericward-noaa** [comment]
I think for the 2-covariate case, you'll want to basically do the same -- but hold each one fixed at 0 (or the mean), and just look at the marginal effects of the other? You could also probably do some cool visualizations in 2D, if you had a function that could take on 2 values of the covariate and generate these probabilities. I'm thinking about a 2 x 2 plot, where each facet was a 2D surface of the transition probability as a function of an X and Y covariate
